### PR TITLE
Synchronization code and notifications

### PR DESCRIPTION
**Pull request by rjzondervan**

### CODE REVIEW

File: Installation/Action/synczaken.action.json [hunk 1]
```diff
@@ -0,0 +1,34 @@
+{
+  "title": "Sync Cases from VrijBRP",
+  "$id": "https://commongateway.nl/action/vrijbrp.synczaken.action.json",
```
Comment (WilcoLouwerse) [inline Installation/Action/synczaken.action.json:3]: maybe use vrijbrp.nl here instead

different domain than mappings from https://github.com/CommonGateway/VrijBRPToZGWBundle/pull/1/files
might be better to be consistent with these references

https://github.com/CommonGateway/VrijBRPToZGWBundle/pull/2/files#r1787422666

File: Installation/Action/vrijbrp.createCaseNotification.action.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{
+  "title": "CreateCaseNotification",
+  "$id": "https://commongateway.nl/action/vrijbrp.createCaseNotification.action.json",
```
Comment (WilcoLouwerse) [inline Installation/Action/vrijbrp.createCaseNotification.action.json:3]: maybe use vrijbrp.nl here instead

File: Installation/Action/vrijbrp.createCaseNotification.action.json [hunk 3]
```diff
@@ -0,0 +1,21 @@
+{
+  "title": "CreateCaseNotification",
+  "$id": "https://commongateway.nl/action/vrijbrp.createCaseNotification.action.json",
+  "$schema": "https://docs.commongateway.nl/schemas/Action.schema.json",
+  "version": "0.0.2",
```
Comment (WilcoLouwerse) [inline Installation/Action/vrijbrp.createCaseNotification.action.json]: version 0.0.1 for new action?

File: Installation/Action/vrijbrp.createStatusNotification.action.json [hunk 4]
```diff
@@ -0,0 +1,21 @@
+{
+  "title": "CreateStatusNotification",
+  "$id": "https://commongateway.nl/action/vrijbrp.createStatusNotification.action.json",
```
Comment (WilcoLouwerse) [inline Installation/Action/vrijbrp.createStatusNotification.action.json:3]: maybe use vrijbrp.nl here instead

File: Installation/Action/vrijbrp.createStatusNotification.action.json [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  "$schema": "https://docs.commongateway.nl/schemas/Action.schema.json",
+  "version": "0.0.2",
```
Comment (WilcoLouwerse) [inline Installation/Action/vrijbrp.createStatusNotification.action.json]: version 0.0.1 for new action?

File: Installation/Mapping/vrijbrp.dossierToZaakDetail.mapping.json [hunk 6]
```diff
@@ -0,0 +1,35 @@
+{
+  "title": "DossierToZaakDetail",
+  "$id": "https://commongateway.nl/mapping/vrijbrp.dossierToZaakDetail.mapping.json",
```
Comment (WilcoLouwerse) [inline Installation/Mapping/vrijbrp.dossierToZaakDetail.mapping.json:3]: maybe use vrijbrp.nl here instead

File: Installation/Mapping/vrijbrp.dossierToZaakDetail.mapping.json [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
...n }}\", \"eigenschap\": {\"_sourceId\": \"BSN_HOOFDBEWONER\", \"naam\": \"BSN_HOOFDBEWONER\"}},{% endif %}{% if previousMunicipality.code|default %}{\"naam\": \"GEMEENTECODE\", \"waarde\": \"{{ previousMunicipality.code }}\", \"eigenschap\": {\"_sourceId\": \"GEMEENTECODE\", \"naam\": \"GEMEENTECODE\"}},{% endif %}{% set index = 0 %}{% for relocator in relocators %}{\"naam\": \"MEEVERHUIZENDE_GEZINSLEDEN.MEEVERHUIZEND_GEZINSLID.{{ index }}.BSN\", \"waarde\": \"{{ relocator.bsn }}\", \"eigenschap\": {\"_sourceId\": \"MEEVERHUIZENDE_GEZINSLEDEN.MEEVERHUIZEND_GEZINSLID.{{ index }}.BSN\", \"naam\": \"MEEVERHUIZENDE_GEZINSLEDEN.MEEVERHUIZEND_GEZINSLID.{{ index }}.BSN\"}},{\"naam\": \"MEEVERHUIZENDE_GEZINSLEDEN.MEEVERHUIZEND_GEZINSLID.{{ index }}.ROL\", \"waarde\": \"{{ relocator.declarationType }}\", \"eigenschap\": {\"_sourceId\": \"MEEVERHUIZENDE_GEZINSLEDEN.MEEVERHUIZEND_GEZINSLID.{{ index }}.ROL\", \"naam\": \"MEEVERHUIZENDE_GEZINSLEDEN.MEEVERHUIZEND_GEZINSLID.{{ index }}.ROL\"}},{% endfor %}{\"naam\": \"AANTAL_PERS_NIEUW_ADRES\", \"waarde\": \"{{ newAddress.numberOfResidents }}\", \"eigenschap\": {\"_sourceId\": \"AANTAL_PERS_NIEUW_ADRES\", \"naam\": \"AANTAL_PERS_NIEUW_ADRES\"}}]"
```
Comment (WilcoLouwerse) [inline Installation/Mapping/vrijbrp.dossierToZaakDetail.mapping.json:19]: This looks weird / not right?

File: src/ActionHandler/NotificationCaseHandler.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    ) {
+
+    }//end __construct()
+
+
+    /**
+     *  This function returns the requered configuration as a [json-schema](https://json-schema.org/) array.
+     *
+     * @throws array a [json-schema](https://json-schema.org/) that this  action should comply to
+     */
```
Comment (WilcoLouwerse) [inline src/ActionHandler/NotificationCaseHandler.php:37]: Missing @return tag in function comment

File: src/ActionHandler/NotificationCaseHandler.php [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    public function getConfiguration(): array
+    {
+        return [
+            '$id'         => 'https://commongateway.nl/ActionHandler/SynchronizationCollectionHandler.ActionHandler.json',
+            '$schema'     => 'https://docs.commongateway.nl/schemas/ActionHandler.schema.json',
+            'title'       => 'SynchronizationCollectionHandler',
+            'description' => '',
+            'required'    => [],
+            'properties'  => [],
```
Comment (WilcoLouwerse) [inline src/ActionHandler/NotificationCaseHandler.php:46]: Missing properties in Handler here

File: src/ActionHandler/NotificationUpdateHandler.php [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    ) {
+
+    }//end __construct()
+
+
+    /**
+     *  This function returns the requered configuration as a [json-schema](https://json-schema.org/) array.
+     *
+     * @throws array a [json-schema](https://json-schema.org/) that this  action should comply to
+     */
```
Comment (WilcoLouwerse) [inline src/ActionHandler/NotificationUpdateHandler.php:38]: Missing @return tag in function comment

File: src/ActionHandler/NotificationUpdateHandler.php [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    public function getConfiguration(): array
+    {
+        return [
+            '$id'         => 'https://commongateway.nl/ActionHandler/SynchronizationCollectionHandler.ActionHandler.json',
+            '$schema'     => 'https://docs.commongateway.nl/schemas/ActionHandler.schema.json',
+            'title'       => 'SynchronizationCollectionHandler',
+            'description' => '',
+            'required'    => [],
+            'properties'  => [],
```
Comment (WilcoLouwerse) [inline src/ActionHandler/NotificationUpdateHandler.php:47]: Missing properties in Handler here

File: src/ActionHandler/SynchronizeCollectionHandler.php [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+ * @license EUPL<github.com/ConductionNL/contactcatalogus/blob/master/LICENSE.md>
+ *
+ * @category ActionHandler
+ */
+class SynchronizeCollectionHandler implements ActionHandlerInterface
+{
+
+
+    /**
+     * @param NewSynchronizationService $syncService
```
Comment (WilcoLouwerse) [inline src/ActionHandler/SynchronizeCollectionHandler.php]: https://github.com/CommonGateway/VrijBRPToZGWBundle/pull/2/files#r1787426617

File: src/Service/VrijBrpService.php [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+    ) {
+
+    }//end __construct()
+
+
+    /**
+     * Set defaults for configuration of synchronize action.
+     *
+     * @param  array $configuration Incoming configuration.
+     * @return array
```
Comment (WilcoLouwerse) [inline src/Service/VrijBrpService.php]: What do you return?

File: src/Service/VrijBrpService.php [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+
+    }//end setVrijBRPDefaults()
+
+
+    /**
+     * Creates a status notification.
+     *
+     * @param  array $data   incoming data.
+     * @param  array $config incoming configuration.
+     * @return array
```
Comment (WilcoLouwerse) [inline src/Service/VrijBrpService.php]: What do you return?

File: src/Service/VrijBrpService.php [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+
+    }//end createStatusNotification()
+
+
+    /**
+     * Specific extension: fetch second source and run mapping.
+     *
+     * @param  ObjectEntity $object Incoming object
+     * @param  array        $array  Serialised incoming object
+     * @return ObjectEntity
```
Comment (WilcoLouwerse) [inline src/Service/VrijBrpService.php]: What do you return?

File: src/Service/VrijBrpService.php [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+
+    }//end extendSync()
+
+
+    /**
+     * Fire notification for case creation. Also extend case with data from detail endpoint.
+     *
+     * @param  array $data   Incoming data
+     * @param  array $config Incoming configuration
+     * @return array
```
Comment (WilcoLouwerse) [inline src/Service/VrijBrpService.php]: What do you return?

**Review by WilcoLouwerse — CHANGES_REQUESTED** [review]
https://github.com/CommonGateway/VrijBRPToZGWBundle/pull/2/files

File: src/ActionHandler/NotificationCaseHandler.php (hunk 9 above)
Comment (rjzondervan) [reply src/ActionHandler/NotificationCaseHandler.php:46]: This action does not take configuration

File: src/ActionHandler/NotificationUpdateHandler.php (hunk 11 above)
Comment (rjzondervan) [reply src/ActionHandler/NotificationUpdateHandler.php:47]: This action does not take configuration as of yet

File: src/Service/NewSynchronizationService.php [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+     * @param array           $sourceObject    The object in the source
+     * @param bool            $unsafe          Unset attributes that are not included in the hydrator array when calling the hydrate function
+     *
+     * @throws GuzzleException
+     * @throws LoaderError
+     * @throws SyntaxError
+     *
+     * @return Synchronization The updated synchronization
+     */
+    public function synchronizeFromSource(Synchronization $synchronization, array $sourceObject=[], bool $unsafe=false): Synchronization
```
Comment (MWest2020) [inline src/Service/NewSynchronizationService.php:76]: Kan naar deze functie gekeken worden? Aardig wat warnings erop? Complexiteit etc

File: Installation/Mapping/vrijbrp.dossierToZaakDetail.mapping.json (hunk 7 above)
Comment (rjzondervan) [reply Installation/Mapping/vrijbrp.dossierToZaakDetail.mapping.json:19]: It is not an optimal mapping, that I grant you, but this is tested against the eigenschappen we have in the current SimXML to ZGW bundle

File: Installation/Action/synczaken.action.json (hunk 1 above)
Comment (rjzondervan) [reply Installation/Action/synczaken.action.json:3]: This is the domain from https://github.com/CommonGateway/VrijBRPToZGWBundle/pull/1/files, that's exactly why I changed the domains to commongateway.nl

File: src/ActionHandler/NotificationCaseHandler.php (hunk 8 above)
Comment (rjzondervan) [reply src/ActionHandler/NotificationCaseHandler.php:37]: This has been updated, but somehow Github does not show (@MWest2020 )

File: src/ActionHandler/NotificationUpdateHandler.php (hunk 10 above)
Comment (rjzondervan) [reply src/ActionHandler/NotificationUpdateHandler.php:38]: This has been updated, but somehow Github does not show (@MWest2020 )

**Review by MWest2020 — APPROVED** [review]

File: src/Service/NewSynchronizationService.php [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+    {
+        $this->synchronizationLogger->info("handleSync for Synchronization with id = {$synchronization->getId()->toString()}");
+
+        // create new object if no object exists
+        if (!$synchronization->getObject()) {
```
Comment (bbrands02) [inline src/Service/NewSynchronizationService.php:81]: Kan !== null

File: src/Service/NewSynchronizationService.php [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+    {
+        $response = $this->callService->call(source: $source, endpoint: $configuration['endpoint'], method: $configuration['method'], config: ['json' => $configuration['body']]);
+
+        $result = $this->callService->decodeResponse(source: $source, response: $response, contentType: ($configuration['content-type'] ?? 'application/json'));
+
+        $resultDot = new Dot($result);
+
+        if ($resultDot->has(keys: $configuration['resultsPath']) === true) {
+            $return = $resultDot->get(key: $configuration['resultsPath']);
+            if ($return instanceof Dot) {
```
Comment (bbrands02) [inline src/Service/NewSynchronizationService.php]: Kan === true

File: src/Service/NewSynchronizationService.php [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+                return $return->jsonSerialize();
+            } else if (is_array($return)) {
```
Comment (bbrands02) [inline src/Service/NewSynchronizationService.php:176]: Kan === true

File: src/Service/VrijBrpService.php [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+     * @param array $data   Incoming data
+     * @param array $config Incoming configuration
+     *
+     * @return array The updated data array.
+     */
+    public function createCaseNotification(array $data, array $config): array
+    {
+        $object = $data['object'];
+
+        if ($object instanceof ObjectEntity) {
```
Comment (bbrands02) [inline src/Service/VrijBrpService.php]: Kan === true

**Review by bbrands02 — APPROVED** [review]

**Pull request merged**